tools/querytee/response_comparator.go

```diff
@@ -186,16 +191,24 @@ func compareScalar(expectedRaw, actualRaw json.RawMessage) error {
 	}, model.SamplePair{
 		Timestamp: actual.Timestamp,
 		Value:     actual.Value,
-	})
+	}, tolerance)
 }
 
-func compareSamplePair(expected, actual model.SamplePair) error {
+func compareSamplePair(expected, actual model.SamplePair, tolerance float64) error {
 	if expected.Timestamp != actual.Timestamp {
 		return fmt.Errorf("expected timestamp %v but got %v", expected.Timestamp, actual.Timestamp)
 	}
-	if expected.Value != actual.Value {
+	if !compareSampleValue(expected.Value, actual.Value, tolerance) {
 		return fmt.Errorf("expected value %s for timestamp %v but got %s", expected.Value, expected.Timestamp, actual.Value)
 	}
 
 	return nil
 }
+
+func compareSampleValue(first, second model.SampleValue, tolerance float64) bool {
+	if tolerance <= 0 {
+		return first == second
```

Comment:
It's not safe to do equality comparisons on floats due to NaNs, use math.Float64bits

Reply:
Thanks for spotting it! The NaN comparison is also wrong when the tolerance is enabled (> 0).

Reply:
:+1: 

We sometimes do this: https://github.com/prometheus/prometheus/blob/e6d7cc5fa43716281bdf5d1cf40415f99a0e9623/storage/series.go#L270 to avoid issues (so use Float compare indeed)